File: 01-java-helloworld/src/main/java/demo/EmulatedNetworkManager.java
package demo;

import com.google.common.collect.ImmutableSet;
import org.protelis.lang.datatype.DeviceUID;
import org.protelis.vm.CodePath;
import org.protelis.vm.NetworkManager;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

/**
 * Emulation of a network manager which uses the shared memory.
 */
public class EmulatedNetworkManager implements NetworkManager {

    private Map<DeviceUID, Map<CodePath, Object>> messages;
    private final DeviceUID uid;
    private ImmutableSet<Device> neighbors;

    /**
     * Constructor method.
     * @param uid the device id
     * @param neighbors the neighbors the device has to send his messages to
     */
    public EmulatedNetworkManager(final DeviceUID uid, final Set<Device> neighbors) {
        this.uid = uid;
        this.setNeighbors(neighbors);
        this.messages = new HashMap<>();
    }

    /**
     * Constructor method for devices with no neighbors.
     * @param uid the device id
     */
    public EmulatedNetworkManager(final DeviceUID uid) {
        this(uid, Collections.emptySet());
    }

    /**
     * Update the network manager with a new set of neighbors.
     * @param neighbors the new neighbors the device has to send his messages to
     */
    public void setNeighbors(final Set<Device> neighbors) {
        this.neighbors = ImmutableSet.copyOf(neighbors);
    }

    /**
     * Receives a message and stores it.
     * @param src the message source
     * @param msg the message content
     */
    private void receiveMessage(final DeviceUID src, final Map<CodePath, Object> msg) {
        messages.put(src, msg);
    }

    /**
     * Called by ProtelisVM read the stored messages.
     * @return the currently stored messages
     */
    @Override
    public Map<DeviceUID, Map<CodePath, Object>> getNeighborState() {
        Map<DeviceUID, Map<CodePath, Object>> t = messages;
        messages = new HashMap<>();
        return t;
    }

    /**
     * Called by ProtelisVM to send a message to the neighbors.
     * @param toSend the message to be sent.
     */
    @Override
    public void shareState(final Map<CodePath, Object> toSend) {
        neighbors.forEach(d -> ((EmulatedNetworkManager) d.getNetworkManager()).receiveMessage(uid, toSend));
    }
}

File: 02-kotlin-helloworld/src/main/kotlin/EmulatedNetworkManager.kt
package demo

import org.protelis.lang.datatype.DeviceUID
import org.protelis.vm.CodePath
import org.protelis.vm.NetworkManager

class EmulatedNetworkManager(private val uid: DeviceUID, var neighbors: Set<Device> = emptySet()) : NetworkManager {

    private var messages: Map<DeviceUID, Map<CodePath, Any>> = emptyMap()

    private fun receiveMessage(src: DeviceUID, msg: Map<CodePath, Any>) {
        messages += Pair(src, msg)
    }

    override fun getNeighborState(): Map<DeviceUID, Map<CodePath, Any>> =
        messages.apply { messages = emptyMap() }

    override fun shareState(toSend: Map<CodePath, Any>) {
        if (toSend.isNotEmpty()) {
            neighbors.forEach { (it.netmgr as EmulatedNetworkManager).receiveMessage(uid, toSend) }
        }
    }
}